src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -274,15 +375,347 @@ TryDropShard(GroupShardPlacement *placement)
+	table_close(pgDistCleanupRecord, NoLock);
+}
+
+
+/*
+ * DeleteCurrentCleanupRecords deletes all failure cleanup records belonging to the
+ * current operation. This is generally used to signal that those objects
+ * have been dropped already by successful completion of the transaction.
+ */
+void
+DeleteMyCleanupOnFailureRecords(void)
+{
+	Relation pgDistCleanupRecord = table_open(DistCleanupRecordRelationId(),
+											  RowExclusiveLock);
+
+	const int scanKeyCount = 2;
+	ScanKeyData scanKey[2];
+	bool indexOK = false;
+
+	/* find failure records belonging to the current operation */
+	OperationId operationId = CurrentOperationId;
+	bool isSuccess = false;
+
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_cleanup_record_operation_id,
+				BTEqualStrategyNumber, F_INT8EQ, UInt64GetDatum(operationId));
+	ScanKeyInit(&scanKey[1], Anum_pg_dist_cleanup_record_is_success,
+				BTEqualStrategyNumber, F_BOOLEQ, BoolGetDatum(isSuccess));
+
+	SysScanDesc scanDescriptor = systable_beginscan(pgDistCleanupRecord,
+													InvalidOid, indexOK,
+													NULL, scanKeyCount, scanKey);
+
+	HeapTuple heapTuple = NULL;
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		simple_heap_delete(pgDistCleanupRecord, &heapTuple->t_self);
+	}
+
+	systable_endscan(scanDescriptor);
+
+	CommandCounterIncrement();
+	table_close(pgDistCleanupRecord, NoLock);
+}
+
+
+/*
+ * StartOperationRequiringCleanup should be called by an operation that wishes to generate
+ * cleanup records.
+ */
+OperationId
+StartOperationRequiringCleanup(void)
+{
+	CurrentOperationId = GetNextOperationId();
+
+	LockOperationId(CurrentOperationId);
+
+	return CurrentOperationId;
+}
+
+
+/*
+ * GetNextOperationId generates a new operation ID using the sequence.
+ */
+static OperationId
+GetNextOperationId(void)
+{
+	RangeVar *sequenceName = makeRangeVar("pg_catalog", "pg_dist_operation_id_seq", -1);
+
+	bool missingOK = false;
+	Oid sequenceId = RangeVarGetRelid(sequenceName, NoLock, missingOK);
+
+	bool checkPermissions = false;
+	return nextval_internal(sequenceId, checkPermissions);
+}
+
+
+/*
+ * LockOperationId takes an exclusive lock on the operation ID to let other
+ * backends know that the operation is active.
+ */
+static void
+LockOperationId(OperationId operationId)
+{
+	LOCKTAG tag;
+	const bool sessionLock = false;
+	const bool dontWait = false;
+	SET_LOCKTAG_CLEANUP_OPERATION_ID(tag, operationId);
+	(void) LockAcquire(&tag, ExclusiveLock, sessionLock, dontWait);
+}
+
+
+/*
+ * IsOperationInProgress checks whether an operation is in progress by
+ * acquiring a share lock on the operation ID, which conflicts with any
+ * transaction that has called LockOperationId.
+ */
+static bool
+IsOperationInProgress(OperationId operationId)
+{
+	LOCKTAG tag;
+	const bool sessionLock = false;
+	const bool dontWait = true;
+	SET_LOCKTAG_CLEANUP_OPERATION_ID(tag, operationId);
+	bool lockAcquired = LockAcquire(&tag, ShareLock, sessionLock, dontWait);
```

Comment:
should we immediately release the lock if lockAcquired is true --so that we don't hold it forever-- ?